### PR TITLE
feature: request logging middleware for express

### PR DESCRIPTION
**Pull request by ofrobots**
Similar to `logging-bunyan`, this PR implements a request correlating middleware for express. This is WIP, but opening now to get early feedback.

### CODE REVIEW

**Comment by ofrobots** [comment]
Tests are expected to fail until a new version of `winston@3` is released that includes the fix https://github.com/winstonjs/winston/pull/1483.

**Comment by ofrobots** [comment]
To spell out the pros/cons more clearly:

- ➖users of logging get extra code and extra dependencies (`@opencensus/propagation-stackdriver`, `on-finished`)
- ➕reduces the overhead of yet another module that needs to be maintained

**Comment by ofrobots** [comment]
I've extracted much of the logic to https://github.com/googleapis/nodejs-logging/pull/248.

**Comment by ofrobots** [comment]
New version of winston is out now. This is ready to review.

File: src/middleware/make-middleware.ts [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as context from '@opencensus/propagation-stackdriver';
+import * as http from 'http';
+
+// TODO: move this code to @google-cloud/logging so that it can be shared
+// betwixt logging-{bunyan,winston}.
```
Comment (JustinBeckwith) [inline src/middleware/make-middleware.ts]: betwixt?

File: src/middleware/make-middleware.ts [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as context from '@opencensus/propagation-stackdriver';
+import * as http from 'http';
+
+// TODO: move this code to @google-cloud/logging so that it can be shared
```
Comment (JustinBeckwith) [inline src/middleware/make-middleware.ts]: Do you have a bug open somewhere to track that?

File: test/middleware/express.ts [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+    assert.strictEqual(loggedEntry!.level, 'miney');
+    assert.strictEqual(loggedEntry!.message, 'catch a tiger by the toe');
+  });
+
+  it('should acquire the projectId and pass it makeMiddleware', async () => {
+    await middleware();
+    assert.strictEqual(passedProjectId, FAKE_PROJECT_ID);
+  });
+
+  it('should throw when projectId is not available');
```
Comment (JustinBeckwith) [inline test/middleware/express.ts]: That's a little weird

**Comment by JustinBeckwith** [comment]
👋 @ofrobots is this ready to go?

**Comment by ofrobots** [comment]
Actually it turns out I still need to do a bit more work on this.

File: package.json [hunk 4]
```diff
@@ -39,7 +39,7 @@
     "samples-test": "cd samples/ && npm link ../ && npm test && cd ../",
     "presystem-test": "npm run pretest",
     "system-test": "mocha build/system-test --timeout 600000",
-    "test": "nyc mocha build/test",
+    "test": "nyc mocha build/test build/test/middleware",
```
[the remark this answers is not in the document]
Comment (ofrobots) [reply package.json]: Done.

File: src/common.ts [hunk 5]
```diff
@@ -77,7 +77,7 @@ function getCurrentTraceFromAgent(): string|null {
 export class LoggingCommon {
   private inspectMetadata: boolean;
   private levels: {[name: string]: number};
-  private stackdriverLog:
+  stackdriverLog:
       types.StackdriverLog;  // TODO: add type for @google-cloud/logging
```
[the remark this answers is not in the document]
Comment (ofrobots) [reply src/common.ts]: Yes, it still makes sense. Logging is still not shipping types. Opened https://github.com/googleapis/nodejs-logging/issues/386.

**Comment by ofrobots** [comment]
Currently blocked on https://github.com/googleapis/nodejs-logging/pull/376 being released.

File: src/common.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (ofrobots) [reply src/common.ts]: I am going to resolve this here. It seems that we are not exporting all the public types at this point. This needs to be addressed in logging.

File: test/middleware/express.ts (hunk 3 above)
Comment (ofrobots) [reply test/middleware/express.ts]: Unweirded.

**Comment by ofrobots** [comment]
Okay, this is ready for review for realz this time.

File: src/middleware/express.ts [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+export async function middleware(options?: MiddlewareOptions) {
+  const defaultOptions = {
+    logName: 'winston_log',
+    level: 'info',
+    levels: winston.config.npm.levels
+  };
+  options = Object.assign({}, defaultOptions, options);
+
+  const loggingWinstonApp = new LoggingWinston(
+      {...options, logName: `${options.logName}_${APP_LOG_SUFFIX}`});
```
Comment (kjin) [inline src/middleware/express.ts]: Would this always replace both the user and default option for `logName`?

**Comment by shaunc869** [comment]
Is there any documentation on how to enable this or does it just magically starting working inside express? Thank you for doing this!

**Comment by ofrobots** [comment]
Hi @shaunc869 this is not publicly exposed at this point unfortunately 😞. This is currently blocked by issues related to TypeScript and our transparent support for both winston2 and winston3 out of the same module: https://github.com/googleapis/nodejs-logging-winston/pull/278. I would suggest subscribing to that issue.

**Comment by shaunc869** [comment]
I just saw the merge come through, looks like we're close on this maybe?

**Comment by ofrobots** [comment]
@shaunc869 yes, this should get released around the end of the month. If you wanted to kick the tires, you can install the module from the master branch and play around: `npm i googleapis/nodejs-logging-winston`. Note that you'll be the first 'real' user of this feature, so this may be rough around  the edges. If you have any feedback, that would be most welcome.

**Comment by shaunc869** [comment]
Sounds great I'll do that tonight any special instructions for this to work with express?

**Comment by ofrobots** [comment]
The readme on the master branch has [some instructions.](https://github.com/googleapis/nodejs-logging-winston#using-as-an-express-middleware-winston-31)

**Comment by shaunc869** [comment]
Dumb question, now my import is failing and I can't see why, can these not be mixed?

`import { LoggingWinston } from '@google-cloud/logging-winston';`

**Comment by DominicKramer** [comment]
The way you are importing `LoggingWinston` should work.  What specific error are you seeing?

**Comment by shaunc869** [comment]
```
"dependencies": {
		"@google-cloud/firestore": "^1.2.0",
		"@google-cloud/language": "^2.1.0",
		"@google-cloud/logging": "^4.5.2",
		"@google-cloud/logging-winston": "googleapis/nodejs-logging-winston",
		"@google-cloud/pubsub": "^0.28.1",
		"@google-cloud/storage": "^2.5.0",
		"@google-cloud/tasks": "0.4.0",
		"@sendgrid/mail": "^6.3.1",
		"absolutify": "^0.1.0",
		"atob": "^2.1.1",
		"axios": "^0.18.0",
		"batchelor": "^2.0.2",
		"bcrypt": "^2.0.0",
		"blob": "^0.0.4",
		"bluebird": "^3.5.1",
		"body-parser": "^1.18.2",
		"cheerio": "^1.0.0-rc.2",
		"cookie-parser": "^1.4.3",
		"cors": "^2.8.1",
		"credit-card-type": "^6.2.0",
		"crypto-js": "^3.1.9-1",
		"csvtojson": "^2.0.8",
		"currency-formatter": "^1.4.2",
		"debug": "^3.1.0",
		"decimal.js": "^9.0.1",
		"dialogflow": "^0.8.0",
		"domain-haven": "^0.0.111",
		"elastic-apm-node": "^2.6.0",
		"elasticsearch": "^15.2.0",
		"es6-promise": "^4.2.2",
		"express": "^4.14.1",
		"express-handlebars": "^3.0.0",
		"express-http-context": "^1.2.1",
		"express-validator": "^4.3.0",
		"ext-name": "^5.0.0",
		"fb": "^2.0.0",
		"fbgraph": "^1.4.4",
		"firebase": "^5.7.1",
		"firebase-admin": "^6.3.0",
		"form-data": "^2.3.2",
		"gmail-api-parse-message": "^2.0.0",
		"google-auth-library": "1.4.0",
		"google-libphonenumber": "^3.1.3",
		"google-search-results-nodejs": "^0.0.1",
		"google-url": "0.0.4",
		"google_my_business": "git+https://github.com/prashant7july/gmb.git",
		"googleapis": "^39.2.0",
		"googleplaces": "^0.6.1",
		"handlebars": "^4.0.5",
		"handlebars-intl": "^1.1.2",
		"https": "^1.0.0",
		"hubspot": "^1.3.0",
		"imagemagick": "^0.1.3",
		"intercom-client": "^2.10.2",
		"locale-currency": "git+https://github.com/procedurallygenerated/locale-currency.git",
		"lodash": "^4.17.5",
		"mailgun-js": "^0.22.0",
		"mailparser": "^2.4.3",
		"moment": "^2.21.0",
		"moment-timezone": "^0.5.21",
		"multer": "^1.4.1",
		"node-adwords": "^201806.0.1",
		"node-linkedin": "^0.5.6",
		"node-quickbooks": "^2.0.24",
		"numeral": "^2.0.6",
		"oauth": "^0.9.15",
		"openid-client": "^2.1.0",
		"path": "^0.12.7",
		"public-ip": "^2.4.0",
		"request": "^2.85.0",
		"request-promise": "^4.2.2",
		"rrule": "^2.6.0",
		"sendgrid": "^5.2.3",
		"sharp": "^0.21.0",
		"simple-oauth2": "^1.5.0",
		"sitemap": "^2.0.1",
		"soap": "^0.25.0",
		"sparkpost": "^2.1.2",
		"stripe": "^6.1.1",
		"twilio": "^3.26.0",
		"twit": "^2.2.9",
		"underscore": "^1.8.3",
		"unfurl.js": "^1.1.8-beta.4",
		"url": "^0.11.0",
		"uuid": "^3.2.1",
		"winston": "3.2.1"
	},
	"devDependencies": {
		"@types/bcrypt": "^2.0.0",
		"@types/cheerio": "^0.22.7",
		"@types/cookie-parser": "^1.4.1",
		"@types/cors": "^2.8.4",
		"@types/credit-card-type": "^5.0.1",
		"@types/currency-formatter": "^1.3.0",
		"@types/dialogflow": "^0.6.3",
		"@types/elasticsearch": "^5.0.29",
		"@types/express-handlebars": "0.0.30",
		"@types/express-session": "^1.15.8",
		"@types/google-cloud__pubsub": "^0.14.1",
		"@types/google-libphonenumber": "^7.4.14",
		"@types/handlebars": "^4.0.37",
		"@types/moment-timezone": "^0.5.9",
		"@types/multer": "^1.3.6",
		"@types/node": "^10.10.1",
		"@types/oauth": "^0.9.0",
		"@types/request": "^2.47.0",
		"@types/simple-oauth2": "^1.1.1",
		"@types/sparkpost": "^2.1.4",
		"@types/stripe": "^5.0.13",
		"@types/twit": "^2.2.12",
		"@types/underscore": "^1.8.8",
		"@types/uuid": "^3.4.4",
		"concurrently": "^3.6.1",
		"cpx": "^1.5.0",
		"request-debug": "^0.2.0",
		"tslint": "^5.8.0",
		"typescript": "^2.9.2"
	}
```

Screen Shot 2019-04-18 at 1 50 32 PM

Before switching to the new library this used to read:

"@google-cloud/logging-winston": "^0.11.1",

and that import worked fine.

**Comment by DominicKramer** [comment]
Either
```js
import * as LoggingWinston from '@google-cloud/logging-winston';

const loggingWinston = new LoggingWinston.LoggingWinston({
	projectId: 'highlevel-backend',
	keyFilename: './lib/highlevel-backend.json',
	serviceContext: {
		service: process.env.GAE_SERVICE,
		version: process.env.GAE_VERSION,
	},
});
```
or
```js
import {LoggingWinston} from '@google-cloud/logging-winston';

const loggingWinston = new LoggingWinston({
	projectId: 'highlevel-backend',
	keyFilename: './lib/highlevel-backend.json',
	serviceContext: {
		service: process.env.GAE_SERVICE,
		version: process.env.GAE_VERSION,
	},
});
```
should work.  Let me know if it doesn't.

**Comment by shaunc869** [comment]
Awesome first one totally worked, now I threw in a test method:

```
export async function logTest(req, res) {
	req.log.info('this is an info log message');
	res.status(200).send('Hello');
}
```

It runs fine, but nothing prints to the console on my localhost, should it?

**Comment by ofrobots** [comment]
@shaunc869 Yeah, that is a problem – the express middleware configures only a stackdriver transport for winston. This is a bug.

However, in order to keep the discussion focussed, I am going to open a new bug rather than continuing to work on this closed issue: https://github.com/googleapis/nodejs-logging-winston/issues/303

**Pull request merged**